velox/exec/tests/MultiFragmentTest.cpp

```diff
@@ -1016,6 +1016,127 @@ TEST_F(MultiFragmentTest, cancelledExchange) {
+    return outputType_;
+  }
+
+  bool* used() const {
+    return used_;
+  }
+
+  const std::vector<core::PlanNodePtr>& sources() const override {
+    static std::vector<core::PlanNodePtr> kEmptySources;
+    return kEmptySources;
+  }
+
+  bool requiresExchangeClient() const override {
+    return true;
+  }
+
+  bool requiresSplits() const override {
+    return true;
+  }
+
+  std::string_view name() const override {
+    return "CustomExchange";
+  }
+
+ private:
+  void addDetails(std::stringstream& stream) const override {
+    // Nothing to add
+  }
+
+  RowTypePtr outputType_;
+  bool* used_;
+};
+
+class TestCustomExchange : public exec::Exchange {
+ public:
+  TestCustomExchange(
+      int32_t operatorId,
+      DriverCtx* FOLLY_NONNULL ctx,
+      const std::shared_ptr<const TestCustomExchangeNode>& customExchangeNode,
+      std::shared_ptr<ExchangeClient> exchangeClient,
+      bool* used)
+      : exec::Exchange(
+            operatorId,
+            ctx,
+            std::make_shared<core::ExchangeNode>(
+                customExchangeNode->id(),
+                customExchangeNode->outputType()),
+            std::move(exchangeClient)),
+        used_(used) {}
+
+  RowVectorPtr getOutput() override {
+    *used_ = true;
+    return exec::Exchange::getOutput();
+  }
+
+ private:
+  bool* used_;
+};
+
+class TestCustomExchangeTranslator : public exec::Operator::PlanNodeTranslator {
+ public:
+  std::unique_ptr<exec::Operator> toOperator(
+      exec::DriverCtx* ctx,
+      int32_t id,
+      const core::PlanNodePtr& node,
+      std::shared_ptr<ExchangeClient> exchangeClient) override {
+    if (auto customExchangeNode =
+            std::dynamic_pointer_cast<const TestCustomExchangeNode>(node)) {
+      return std::make_unique<TestCustomExchange>(
+          id,
+          ctx,
+          customExchangeNode,
+          std::move(exchangeClient),
+          customExchangeNode->used());
+    }
+    return nullptr;
+  }
+};
+
+TEST_F(MultiFragmentTest, customPlanNodeWithExchangeClient) {
+  setupSources(5, 100);
+  Operator::registerOperator(std::make_unique<TestCustomExchangeTranslator>());
+  auto leafTaskId = makeTaskId("leaf", 0);
+  auto leafPlan =
+      PlanBuilder().values(vectors_).partitionedOutput({}, 1).planNode();
+  auto leafTask = makeTask(leafTaskId, leafPlan, 0);
+  Task::start(leafTask, 1);
+
+  bool customOperatorUsed = false;
+  CursorParameters params;
+  params.maxDrivers = 1;
+  params.planNode =
+      PlanBuilder()
+          .addNode([&leafPlan, &customOperatorUsed](
+                       std::string id, core::PlanNodePtr /* input */) {
+            return std::make_shared<TestCustomExchangeNode>(
+                id, leafPlan->outputType(), &customOperatorUsed);
+          })
+          .planNode();
+
+  auto cursor = std::make_unique<TaskCursor>(params);
+  auto task = cursor->task();
+  EXPECT_FALSE(customOperatorUsed);
+  addRemoteSplits(task, {leafTaskId});
+  while (cursor->moveNext()) {
```

Comment:
It is not very clear to me how does this test verify custom exchange logic. Would you explain? Would it be possible to use AssertQueryBuilder and check the runtime stats to confirm there is custom exchange operator instead of "used" logic?